### PR TITLE
[SPARK-26830][SQL][R] Vectorized R dapply() implementation

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR targets to add vectorized `dapply()` in R, Arrow optimization.

This can be tested as below:

```bash
$ ./bin/sparkR --conf spark.sql.execution.arrow.enabled=true
```

```r
df <- createDataFrame(mtcars)
collect(dapply(df, function(rdf) { data.frame(rdf$gear + 1) }, structType("gear double")))
```

### Requirements
  - R 3.5.x 
  - Arrow package 0.12+
    ```bash
    Rscript -e 'remotes::install_github("apache/arrow@apache-arrow-0.12.0", subdir = "r")'
    ```

**Note:** currently, Arrow R package is not in CRAN. Please take a look at ARROW-3204.
**Note:** currently, Arrow R package seems not supporting Windows. Please take a look at ARROW-3204.

### Benchmarks

**Shall**

```bash
sync && sudo purge
./bin/sparkR --conf spark.sql.execution.arrow.enabled=false --driver-memory 4g
```

```bash
sync && sudo purge
./bin/sparkR --conf spark.sql.execution.arrow.enabled=true --driver-memory 4g
```

**R code**

```r
rdf <- read.csv("500000.csv")
df <- cache(createDataFrame(rdf))
count(df)


test <- function() {
  options(digits.secs = 6) # milliseconds
  start.time <- Sys.time()
  count(cache(dapply(df, function(rdf) { rdf }, schema(df))))
  end.time <- Sys.time()
  time.taken <- end.time - start.time
  print(time.taken)
}

test()
```

**Data (350 MB):**

```r
object.size(read.csv("500000.csv"))
350379504 bytes
```

"500000 Records"  http://eforexcel.com/wp/downloads-16-sample-csv-files-data-sets-for-testing/

**Results**

```
Time difference of 13.42037 mins
```

```
Time difference of 30.64156 secs
```

The performance improvement was around **2627%**.

### Limitations

- For now, Arrow optimization with R does not support when the data is `raw`, and when user explicitly gives float type in the schema. They produce corrupt values.

- Due to ARROW-4512, it cannot send and receive batch by batch. It has to send all batches in Arrow stream format at once. It needs improvement later.

## How was this patch tested?

Unit tests were added, and manually tested.

### CODE REVIEW

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -1437,6 +1437,29 @@ dapplyInternal <- function(x, func, schema) {
     schema <- structType(schema)
   }
 
+  arrowEnabled <- sparkR.conf("spark.sql.execution.arrow.enabled")[[1]] == "true"
```
Comment (HyukjinKwon) [inline R/pkg/R/DataFrame.R]: @felixcheung, I will deduplicate this checker across all R optimization after this PR if that's fine to you too. I was trying to work on multiple features so wanted to avoid conflicts.

**Comment by HyukjinKwon** [comment]
cc @BryanCutler, @viirya, @felixcheung, @icexelloss, @rxin, @gatorsmile, @shivaram, @falaki, @yanboliang

**Comment by HyukjinKwon** [comment]
FWIW, my todos on my current plate is:

- Set socket timeout consistently
- Deduplicate schema checking
- Timestamp / date support in regaulr gapply (without Arrow)
- Refactor PythonRunner
- Refactor Auth
- Arrow R CRAN fix (if Arrow is on CRAN).

**Comment by HyukjinKwon** [comment]
cc @cloud-fan too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 2]
```diff
@@ -193,6 +194,87 @@ case class MapPartitionsExec(
...
+      // 4.                                    Computes R native function on the data frame
+      // 5.                                    Converts the R data frame to Arrow record batches
+      // 6. Columnar batches         <-------- Arrow record batches
+      // 7. Each row from each batch
+      //
+      // Note that, unlike Python vectorization implementation, R side sends Arrow formatted
+      // binary in a batch due to the limitation of R API. See also ARROW-4512.
+      val columnarBatchIter = runner.compute(batchIter, -1)
+      val outputProject = UnsafeProjection.create(output, output)
+      new Iterator[InternalRow] {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: This is almost as is from Scalar Pandas's iterator.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 3]
```diff
@@ -193,6 +194,87 @@ case class MapPartitionsExec(
...
+        if (batchSize > 0) new BatchIterator(inputIter, batchSize) else Iterator(inputIter)
+
+      val runner = new ArrowRRunner(func, packageNames, broadcastVars, inputSchema,
+        SQLConf.get.sessionLocalTimeZone, RRunnerModes.DATAFRAME_DAPPLY)
+
+      // The communication mechanism is as follows:
+      //
+      //    JVM side                           R side
+      //
+      // 1. Internal rows            --------> Arrow record batches
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: One difference comparing to Scalar Pandas UDF here is, it buffers Arrow batches (namely R Data Frames at R side) and applies the function at once. This difference is because dapply explicitly meantions that the function is applied to each partition:

https://github.com/apache/spark/blob/2f6e88fecb455a02c4c08c41290e2f338e979543/R/pkg/R/DataFrame.R#L1461

FYI, @mengxr per SPARK-26412

File: sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala [hunk 4]
```diff
@@ -55,13 +56,32 @@ class ArrowRRunner(
     numPartitions = -1,
     isDataFrame = true,
     schema.fieldNames,
-    RRunnerModes.DATAFRAME_GAPPLY) {
+    mode) {
+
+  // TODO: it needs to refactor to share the same code with RRunner, and have separate
+  // ArrowRRunners.
+  private val getNextBatch = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala]: I plan to clean this whole ArrowRRunner to deduplicate codes with RRunner after this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 5]
```diff
@@ -193,6 +194,87 @@ case class MapPartitionsExec(
...
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    child.execute().mapPartitionsInternal { iter =>
+      // Hyukjin: seems to convert from Arrow batches to Arrow batches directly here
+      // causes segmentation fault intermittently. For instance, when child is directly
+      // originated from Arrow batches. Input projection was added to avoid it based on
+      // my speculation (and it does not happen anymore). It is potentially an issue in Arrow
+      // optimization in my humble opinion.
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: I don't have an explicit evidence for this.

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 6]
```diff
@@ -3300,6 +3300,101 @@ test_that("dapplyCollect() on DataFrame with a binary column", {
...
+  finally = {
+    # Resetting the conf back to default value
+    callJMethod(conf, "set", "spark.sql.execution.arrow.enabled", arrowEnabled)
+  })
+})
+
+test_that("dapply() Arrow optimization - type specification (date and timestamp)", {
+  skip_if_not_installed("arrow")
+  rdf <- data.frame(list(list(a = as.Date("1990-02-24"),
+                              b = as.POSIXct("1990-02-24 12:34:56"))))
```
Comment (HyukjinKwon) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: I plan to add dates and timestamp support for dapply as well so that we can deduplicate tests.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala (hunk 5 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: This only happens when the input batchs are pruned by https://github.com/apache/spark/pull/23810. The root cause seems when the input columns are less than the schema set to Arrow writer.

I am going to work around by the same way. In that way, it's guaranteed that the input columns are not being pruned comparing to Arrow writer.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: Seems this happens only when the input is pruned but Arrow writer expects full schema. I will follow the same approach of https://github.com/apache/spark/pull/23810 for `gapply()`.
Therefore, the input will be guaranteed to not be pruned.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

File: R/pkg/R/DataFrame.R (hunk 1 above)
Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R]: I filed a JIRA at https://issues.apache.org/jira/browse/SPARK-26920

**Comment by HyukjinKwon** [comment]
I am sure this is ready for a review. Many codes are similar with both Scalar vectorized UDF and vectorized gapply(). Should be quite safe to go.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala (hunk 4 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala]: I filed a JIRA https://issues.apache.org/jira/browse/SPARK-26923

**Comment by HyukjinKwon** [comment]
gentle ping

File: R/pkg/R/DataFrame.R [hunk 7]
```diff
@@ -1437,6 +1437,29 @@ dapplyInternal <- function(x, func, schema) {
...
+    if (inherits(schema, "structType")) {
+      if (any(sapply(schema$fields(), function(x) x$dataType.toString() == "FloatType"))) {
+        stop("Arrow optimization with dapply do not support FloatType yet.")
+      }
+      if (any(sapply(schema$fields(), function(x) x$dataType.toString() == "BinaryType"))) {
+        stop("Arrow optimization with dapply do not support BinaryType yet.")
+      }
+    } else if (is.null(schema)) {
+      stop(paste0("Arrow optimization does not support 'dapplyCollect' yet. Please use ",
+                  "'collect' and 'dapply' APIs instead."))
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: this isn't quite the same though, for example, dapply will need the user to pass in a schema. should we suggest turning off arrow then use dapplyCollect? I'm actually not sure - maybe suggest both in the message?

**Review by felixcheung — COMMENTED** [review]
I reviewed the R side

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 8]
```diff
@@ -3300,6 +3300,104 @@ test_that("dapplyCollect() on DataFrame with a binary column", {
...
+  })
+
+  callJMethod(conf, "set", "spark.sql.execution.arrow.enabled", "true")
+  tryCatch({
+    ret <- dapply(df,
+                  function(rdf) {
+                    stopifnot(class(rdf) == "data.frame")
+                    # mtcars' hp is more then 50.
+                    stopifnot(all(rdf$hp > 50))
+                     rdf
```
Comment (BryanCutler) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: is the indentation off here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala [hunk 9]
```diff
@@ -55,13 +56,32 @@ class ArrowRRunner(
...
+      }
+    }
+  }
 
   protected override def writeData(
       dataOut: DataOutputStream,
       printOut: PrintStream,
-      iter: Iterator[_]): Unit = if (iter.hasNext) {
-    val inputIterator = iter.asInstanceOf[Iterator[(Array[Byte], Iterator[InternalRow])]]
+      inputIterator: Iterator[_]): Unit = if (inputIterator.hasNext) {
```
Comment (BryanCutler) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala]: Is it possible for `inputIterator.hasNext` to be false, like from an empty partition, and is there a test for this? It might be better to send an empty Arrow stream in this case, but I'm not entirely sure what the R worker is expecting.

**Review by BryanCutler — COMMENTED** [review]
Thanks for doing this @HyukjinKwon !  Looks pretty good from what I can tell. Possibly adding a few more tests here or in a followup might be good for some edge cases.

File: R/pkg/tests/fulltests/test_sparkSQL.R (hunk 8 above)
Comment (HyukjinKwon) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: Do'h

File: R/pkg/R/DataFrame.R (hunk 7 above)
Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R]: Yes, actually that's why I didn't implemented yet .. as we discussed in the JIRA that targets `gapplyCollect`.

I am not very sure about that too - I thought promoting Arrow optimization might be a good idea although indeed it's not exactly same since `dapplyCollect` can omit schema whereas `dapply` requires it.

Since the message says `Arrow optimization does not support ...`, I roughly guess users would disable it too if they cannot use `collect` and `dapply` later.

I currently think it's okay only to say Arrow optimization as an alternative but I am willing to note to disable Arrow optimization to use `dapplyCollect` if you feel kind of strongly about it.

Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R]: Hm, we could revisit this anyway when I write documentation about this at SPARK-26924

File: sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala (hunk 9 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala]: Oh, yea. I actually manually tested them. I'll fix the tests to check it as well.

Actually, I noticed R worker works a bit differently.

If the iterator is empty, it sends `0` first and then the other data is not used. (see `isEmpty <- SparkR:::readInt(inputCon)` at `worker.R` and,

```scala
          if (!iter.hasNext) {
            dataOut.writeInt(0)
          } else {
            dataOut.writeInt(1)
          }
```

at `RRunner.scala`).

So, virtually it doesn't matter if we send the data or not. I just simply decided to ignore this since it's already ignored at R's worker side.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala]: and .. I target to add more type specification test (for instance, map type and struct type are not supported).

I tried to only switch data transfer logic so checking empty rows, non-empty rows, columns, type specifications might be okay enough. (BTW, existing dapply itself is a bit flaky with some holes IMHO so it's a bit tricky to compare both).

**Comment by HyukjinKwon** [comment]
retest this please

File: R/pkg/R/DataFrame.R (hunk 7 above)
Comment (felixcheung) [reply R/pkg/R/DataFrame.R]: let's aim to fix ?applyCollect anyway - I feel because of simpler usage it might be more popular to use vs. ?apply versions

Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R]: Okie. Let me push some changes soon

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala [hunk 10]
```diff
@@ -34,7 +34,7 @@ import org.apache.spark.sql.types.StructType
  * This is necessary because sometimes we cannot hold reference of input rows
  * because the some input rows are mutable and can be reused.
  */
-private class BatchIterator[T](iter: Iterator[T], batchSize: Int)
+private[spark] class BatchIterator[T](iter: Iterator[T], batchSize: Int)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala]: BTW this is also under python package. Let me try to pull it out this one as well in followup.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala (hunk 9 above)
Comment (BryanCutler) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala]: Sounds good to me, thanks!

**Review by BryanCutler — APPROVED** [review]
LGTM from my end

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

I am going to resolve the followups one by one.

**Comment by HyukjinKwon** [comment]
Thanks, @felixcheung and @BryanCutler
